File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.js
function solution(cards) {
    let answers = [];
    let check = new Array(cards.length).fill(false);
    
    for(let i = 0; i < cards.length; i++) {
        let cur = cards[i];
        let cnt = 0;
        while(!check[cur - 1]) {
            check[cur - 1] = true;
            cur = cards[cur - 1];
            cnt++;
        }
        
        if(cnt)
            answers.push(cnt);
    }
    
    answers.sort((a, b) => b - a);
    if(answers.length < 2)
        return 0;
    
    return answers[0] * answers[1];
}

File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool Comp(int a, int b) {
    return a > b;
}

int solution(vector<int> cards) {
    vector<int> answers;
    vector<bool> check(cards.size(), false);
    
    for(int i = 0; i < cards.size(); i++) {
        int cur = cards[i];
        int cnt = 0;
        while(!check[cur - 1]) {
            check[cur - 1] = true;
            cur = cards[cur - 1];
            cnt++;
        }
        
        if(cnt)
            answers.push_back(cnt);
    }
    
    sort(answers.begin(), answers.end(), Comp);
    if(answers.size() > 1)
        return answers[0] * answers[1];
    
    return 0;
}